azure-iot-device/azure/iot/device/iothub/abstract_clients.py

```diff
@@ -67,6 +67,24 @@ def create_from_shared_access_signature(cls, sas_token):
         pipeline = IoTHubPipeline(authentication_provider)
         return cls(pipeline)
 
+    @classmethod
+    def create_from_x509_certificate(cls, hostname, x509, device_id, module_id=None):
```

Comment:
This should also be two separate APIs, one for device, and one for module, given that the signatures are different.

Reply:
Yes, but not only because the signatures are different.  They are also intended to create different objects -- they should be on AbstractIoTHubModuleClient and AbstractIoTHubDeviceClient.  Otherwise, we're giving implicit permission to add a module ID when creating a device client or skip the module ID when creating a module client.

---
In reply to: [301326329](https://github.com/Azure/azure-iot-sdk-python-preview/pull/129#discussion_r301326329) [](ancestors = 301326329)